CODING_CONVENTIONS.md

```diff
@@ -43,6 +43,39 @@
 * You can use [uncrustify](http://uncrustify.sourceforge.net/) with the provided
   option files: https://github.com/RIOT-OS/RIOT/blob/master/uncrustify-riot.cfg
 
+## Standard Compliance
+
+Using extensions to the C standard in general decreases portability and
+maintainability: The former because porting RIOT to platforms for which limited
+compiler options are available becomes more difficult when compiler specific
+extensions are used. The latter because extensions are often not as clearly
+defined as standard C, not as well known within the C development community,
+and have fewer resources to look up.
+
+There are a number of cases in which using extensions cannot be avoided, or
+would not be maintainable. For these cases, an exception can be made. A list
+of recognised exceptions where we can (or even must) rely on extensions include:
+
+- Use of `__attribute__((packed))` is allowed for serialization and
+  de-serialization and only there. Ideally, it should not be used in public
+  APIs and types.
+- Code specific to MCU families may use extensions commonly used in this domain,
+  such as inline assembly (e.g. as needed for context swapping), special
+  function attributes (e.g. as needed for IRQ vector entries on some MCUs),
+  etc. Code should still prefer standard compliance when there is no significant
+  downside to it compared to using the extension.
+- `#include_next` may be used when system headers need to be extended.
+- Function attributes for which a wrapper exists in `compiler_hints.h` may be
+  used using that wrapper. These wrappers either unlock additional optimization
+  (such as `NORETURN` or `PURE`) or influence warnings (such as `MAYBE_UNUSED`)
+  produced by the compiler and can simply be replaced by an empty token for
+  compilers that do not support them.
+- `__attribute__((used))`, `__attribute__((section("...")))`,
+  `__attribute__((weak))`, and `__attribute__((alias("...")))`
```

Comment:
Notably missing but used in the code base:

- `__attribute__((always_inline))`: This should be IMO a wrapper in `compiler_hints.h`. A compiler not supporting this will generate less efficient code (assuming this is actually sensibly used), but correct code, if this is ignored
- `__attribute__((align(N)))`/__attribute__((aligned(N)))`: There is now `<stdalign.h>` that provides `alignas()` and the `_Alignas()` since C11, so no need to use an extension for this.